### PR TITLE
Doing Dumb Stuff with Postgres

### PR DESCRIPTION
**Pull request by RyanMacG**

### CODE REVIEW

File: posts/elixir_my_second_programming_language_love.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Elixir: My Second Programming Language Love
```
Comment (sowasred2012) [inline posts/elixir_my_second_programming_language_love.md]: You sneaking in another post here?

File: posts/dumb_things_with_postgres.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+# PUT DIAGRAM/screenshot ABOVE
+
+Doing this provides us a couple of advantages, with caching this can be noticeably faster than joining the 3 seperate tables and provides us a clean datastructure that we don't have to manipulate or `pluck` from to get the thing we care about, in this case the tag IDs. It also allows us to easily execute some very basic SQL to find articles that have the same tags as the current one being shown to the site user.
+
+```ruby
+class Article < ActiveRecord::Base
+  # all the other logic
+
+  def associated_articles
+    Article.where(id != self.id).where("tags = ANY(#{self.tags})")
```
Comment (zuren) [inline posts/dumb_things_with_postgres.md]: `Article.where(id != self.id)` - that's not a thing is it?

Comment (RyanMacG) [inline posts/dumb_things_with_postgres.md]: should probably be `where.not` going to test all the code before I wrap this

Comment (zuren) [inline posts/dumb_things_with_postgres.md]: Good on you, never run the stuff I write. always scared of blogging broken code.

Yeah `where.not(id: self.id)` I guess.

File: posts/dumb_things_with_postgres.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+```ruby
+  change_column :products, :product_category_id, 'integer[] USING ARRAY[product_category_id]::INTEGER[]', array: true, null: false, default: []
+```
+This line takes our existing foreign key column and converts it to and array type, the SQL being executed recasts the existing data in the correct format so `4` would become `{4}`. The curly braces are how Postgres denotes an array, when returned by ActiveRecord the data would be represented as a standard array.
+
+```ruby
+  add_index :products, :product_category_ids, using: 'gin'
+```
+This is clearly an index, the important part however is that it's using "GIN" as it's index type. GIN is lossless versus GiST, the other index type available, which is lossy introducting the chance for incorrect results slightly. In practice it probably won't make a huge amount of difference to you which one you use but if you want to read more on these types I'd recommend the [Postgres documentation](https://www.postgresql.org/docs/current/static/textsearch-indexes.html). One thing to watch out for is that GIN indexes can have long build times but there are ways to mitigate that.
```
Comment (adrianclay) [inline posts/dumb_things_with_postgres.md]: I'd mention it is a full-text index type somewhere, I didn't realise until following the linkaroo.

File: posts/elixir_my_second_programming_language_love.md (hunk 1 above)
Comment (RyanMacG) [inline posts/elixir_my_second_programming_language_love.md]: Haha no need to delete that, was throwing around a bunch of ideas

File: posts/dumb_things_with_postgres.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+It's interesting to note that by choosing not to sort this array and restore it would allow the potential of using `SELECT unnest(arr[1:array_length(arr, 1)][1]) as id from data` or similar to get the first item to create a reversible migration. This is only advisable as a thought exercise and I'd recommend you _never_ do that.
+
+The first snag we'll hit here is that we no longer have support for any of the Rails nicities around associations like `has_many` you can solve this by using code similar to the example above;
+
+```ruby
+  def product_categories
+    ProductCategories.where(id: self.product_category_ids)
```
Comment (lukemorton) [inline posts/dumb_things_with_postgres.md]: No need for `self.`

File: posts/dumb_things_with_postgres.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+![Array tags](dumb_things_with_postgres/Screenshot from 2017-05-22 09-09-26.png)
+
+Doing this provides us a couple of advantages, with caching this can be noticeably faster than joining the 3 seperate tables and provides us a clean datastructure that we don't have to manipulate or `pluck` from to get the thing we care about, in this case the tags. It also allows us to easily execute some very basic SQL to find articles that have the same tags as the current one being shown to the site user.
+
+```ruby
+class Article < ActiveRecord::Base
+  # all the other logic
+
+  def associated_articles
+    Article.where.not(id: self.id).where('tags && ARRAY[?]', self.tags)
```
Comment (lukemorton) [inline posts/dumb_things_with_postgres.md]: You only need `self.` when assigning values within an ActiveRecord model.

``` ruby
def set_defaults
  self.age = 44
  self.tags ||= []
end

def full_name
  "#{first_name} #{last_name}"
end
```

Comment (RyanMacG) [inline posts/dumb_things_with_postgres.md]: All gone now

**Pull request merged**